components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAAuthenticationUtil.java

```diff
@@ -0,0 +1,268 @@
+
+        String carbonHome = System.getProperty(CarbonBaseConstants.CARBON_HOME);
+
+        // set the krb5.conf file path if not set by the system property already
+        if (IdentityUtil.isBlank(kerberosFilePath)) {
+            kerberosFilePath =
+                    Paths.get(carbonHome, "repository", "conf", "identity", IWAConstants.KERBEROS_CONF_FILE_NAME)
+                            .toString();
+            System.setProperty(IWAConstants.KERBEROS_CONFIG_FILE, kerberosFilePath);
+        }
+
+        // set jaas.conf file path if not set by the system property already
+        if (IdentityUtil.isBlank(jaasConfigPath)) {
+            jaasConfigPath = Paths.get(carbonHome, "repository", "conf", "identity", IWAConstants.JAAS_CONF_FILE_NAME)
+                    .toString();
+            System.setProperty(IWAConstants.JAAS_CONFIG_FILE, jaasConfigPath);
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Kerberos config file path set : " + kerberosFilePath + " ,JAAS config file path set : "
+                    + jaasConfigPath);
+        }
+
+    }
+
+    /**
+     * Create server credential using SPN
+     *
+     * @param callbackHandler username password callback handler
+     * @throws PrivilegedActionException
+     * @throws LoginException
+     */
+    private static GSSCredential createServerCredentials(CallbackHandler callbackHandler)
+            throws PrivilegedActionException, LoginException {
+        // authenticate to the AD (Kerberos Server)
+        LoginContext loginContext = new LoginContext(IWAConstants.SERVER, callbackHandler);
+        loginContext.login();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Pre-authentication successful for with Kerberos Server.");
+        }
+        // create server credentials from pre authentication with the AD
+        return createCredentialsForSubject(loginContext.getSubject());
+    }
+
+
+    /**
+     * Create GSSCredential as Subject
+     *
+     * @param subject login context subject
+     * @return GSSCredential
+     * @throws PrivilegedActionException
+     */
+    private static GSSCredential createCredentialsForSubject(final Subject subject) throws PrivilegedActionException {
+        final PrivilegedExceptionAction<GSSCredential> action =
+                new PrivilegedExceptionAction<GSSCredential>() {
+                    public GSSCredential run() throws GSSException {
+                        return gssManager.createCredential(null, GSSCredential.INDEFINITE_LIFETIME,
+                                IWAServiceDataHolder.getSpnegoOid(), GSSCredential.ACCEPT_ONLY);
+                    }
+                };
+
+        if (log.isDebugEnabled()) {
+            Set<Principal> principals = subject.getPrincipals();
+            String principalName = null;
+            if (principals != null) {
+                principalName = principals.toString();

```

Comment:
This set can contain only one Principal object at maximum. So the toString of this set is the toString of a single Principal object.